spec/outputs/dynatrace_spec.rb

```diff
@@ -19,133 +19,97 @@
-    end
-
-    let(:last_request) { TestApp.last_request }
-    let(:body) { last_request.body.read }
-    let(:content_type) { last_request.env['CONTENT_TYPE'] }
-    let(:authorization) { last_request.env['HTTP_AUTHORIZATION'] }
-
-    let(:expected_body) { "#{LogStash::Json.dump([event])}\n" }
-    let(:expected_content_type) { 'application/json; charset=utf-8' }
-    let(:expected_authorization) { "Api-Token #{key}" }
-
-    it 'should not log a failure' do
-      expect(subject).not_to have_received(:log_failure).with(any_args)
+  context 'server response success' do
+    it 'sends events' do
+      allow(subject).to receive(:send) do |req|
+        body = JSON.parse(req.body)
+        expect(body.length).to eql(2)
+        expect(body[0]['message']).to eql('message 1')
+        expect(body[0]['@timestamp']).to eql('2021-06-25T15:46:45.693Z')
+        expect(body[1]['message']).to eql('message 2')
+        expect(body[1]['@timestamp']).to eql('2021-06-25T15:46:46.693Z')
+        Net::HTTPOK.new "1.1", "200", "OK"
+      end
+      subject.multi_receive(events)
+      expect(subject).to have_received(:send)
     end
 
-    it 'should receive the request' do
-      expect(last_request).to be_truthy
+    it 'includes authorization header' do
+      allow(subject).to receive(:send) do |req|
+        expect(req['Authorization']).to eql("Api-Token #{key}")
+        Net::HTTPOK.new "1.1", "200", "OK"
+      end
+      subject.multi_receive(events)
+      expect(subject).to have_received(:send)
     end
 
-    it 'should receive the event as a hash' do
-      expect(body).to eql(expected_body)
+    it 'includes content type header' do
+      allow(subject).to receive(:send) do |req|
+        expect(req['Content-Type']).to eql('application/json; charset=utf-8')
+        Net::HTTPOK.new "1.1", "200", "OK"
+      end
+      subject.multi_receive(events)
+      expect(subject).to have_received(:send)
     end
 
-    it 'should have the correct content type' do
-      expect(content_type).to eql(expected_content_type)
+    it 'includes user agent' do
+      allow(subject).to receive(:send) do |req|
+        expect(req['User-Agent']).to eql('logstash-output-dynatrace v1.2.3')
+        Net::HTTPOK.new "1.1", "200", "OK"
+      end
+      subject.multi_receive(events)
+      expect(subject).to have_received(:send)
     end
+  end
 
-    it 'should have the correct authorization' do
-      expect(authorization).to eql(expected_authorization)
+  context 'with bad client request' do
+    it 'does not retry on 404' do
+      allow(subject).to receive(:send) { Net::HTTPNotFound.new "1.1", "404", "Not Found" }
+      subject.multi_receive(events)
+      expect(subject).to have_received(:send).once
     end
   end
 
-  context 'with failing requests' do
-    let(:url) { "http://localhost:#{port}/bad" }
+  context 'with server error' do
+    it 'retries 5 times with exponential backoff' do
+      allow(subject).to receive(:sleep)
```

Comment:
Will this make the sleep call a no-op so the test doesn't take unnecessarily long?

Reply:
yes